File: MachineLearningAlgorithms/3-supervisedML/2-classification/3-supportVectorMachine/support_vector_machine.R
setwd("~/gitMachineLearning/3-supervisedML/2-classification/")

input_df <- read.csv("Social_Network_Ads.csv", stringsAsFactors = FALSE)

input_df <- input_df[2:5]

#Categorical casting
input_df$Gender <- as.numeric(factor(input_df$Gender, 
                                     levels = c('Male', 'Female'),
                                     labels = c(1,2)))
input_df$Purchased <- as.factor(input_df$Purchased)

#Splitting in train and test set
library(caTools)
ind <- sample.split(input_df$Purchased, SplitRatio = .8)
train_set <- input_df[ind == TRUE, ]
test_set <- input_df[ind == FALSE, ]

#Feature scaling
train_set[-4] <- scale(train_set[-4])
test_set[-4] <- scale(test_set[-4])

#model building
#install.packages("e1071")
library(e1071)
model <- svm(formula = Purchased ~ . , 
             data = train_set,
             type = 'C-classification', # Default type for classification is C-classification.
             kernel = "linear")
y_pred <- predict(model, newdata = test_set )

#Evaluation by confusion matrix
cm <- table(y_pred, test_set$Purchased)

File: MachineLearningAlgorithms/3-supervisedML/2-classification/3-supportVectorMachine/supportVectorMachine.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#Reading dataset
data = pd.read_csv("/Users/z001l92/gitMachineLearning/3-supervisedML/2-classification/Social_Network_Ads.csv")
X = data.iloc[:, 1:4].values
Y = data.iloc[:, 4].values

#Categorical casting
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
X_lbl = LabelEncoder()
X[:, 0] = X_lbl.fit_transform(X[:, 0])
X_enc = OneHotEncoder(categorical_features=[0])
X = X_enc.fit_transform(X).toarray()


#Train Test split
from sklearn.cross_validation import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = .25)


#Feature scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)


#Model Building
from sklearn.svm import SVC
model = SVC(kernel = 'linear')
model.fit(X_train, Y_train)

#Predicting output
Y_pred = model.predict(X_test)

#Evaluating results
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, Y_pred)
print(cm)
